### ISSUE TITLE
IllegalArgumentException on incremental build in IntelliJ 2020.3

### ISSUE THREAD

**Issue by MichalStehlikCz** [typetools/checker-framework#4068]
When using checkerframework and incremental build is enabled in IntelliJ 2020.3, build fails with IllegalArgumentException

Based on discussion in https://youtrack.jetbrains.com/issue/IDEA-256707, JetBrains changed behaviour of IntelliJ - they now wrap com.sun.tools.javac.processing.JavacProcessingEnvironment in dynamic proxy, resulting in following exception when environment is passed to JavacTask in AbstractTypeProcessor.init()

2020-12-23 15:33:10,479 [   2252]   INFO - s.incremental.java.JavaBuilder - java.lang.IllegalArgumentException
java.lang.RuntimeException: java.lang.IllegalArgumentException
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:163)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:231)
	at org.jetbrains.jps.incremental.java.JavaBuilder.compileJava(JavaBuilder.java:498)
	at org.jetbrains.jps.incremental.java.JavaBuilder.compile(JavaBuilder.java:350)
	at org.jetbrains.jps.incremental.java.JavaBuilder.doBuild(JavaBuilder.java:274)
	at org.jetbrains.jps.incremental.java.JavaBuilder.build(JavaBuilder.java:228)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runModuleLevelBuilders(IncProjectBuilder.java:1441)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuildersForChunk(IncProjectBuilder.java:1100)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildTargetsChunk(IncProjectBuilder.java:1224)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunkIfAffected(IncProjectBuilder.java:1066)
	at org.jetbrains.jps.incremental.IncProjectBuilder.buildChunks(IncProjectBuilder.java:832)
	at org.jetbrains.jps.incremental.IncProjectBuilder.runBuild(IncProjectBuilder.java:419)
	at org.jetbrains.jps.incremental.IncProjectBuilder.build(IncProjectBuilder.java:183)
	at org.jetbrains.jps.cmdline.BuildRunner.runBuild(BuildRunner.java:132)
	at org.jetbrains.jps.cmdline.BuildSession.runBuild(BuildSession.java:302)
	at org.jetbrains.jps.cmdline.BuildSession.run(BuildSession.java:132)
	at org.jetbrains.jps.cmdline.BuildMain$MyMessageHandler.lambda$channelRead0$0(BuildMain.java:219)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalArgumentException
	at jdk.compiler/com.sun.source.util.JavacTask.instance(JavacTask.java:65)
	at org.checkerframework.javacutil.AbstractTypeProcessor.init(AbstractTypeProcessor.java:97)
	at org.checkerframework.framework.source.SourceChecker.init(SourceChecker.java:502)
	at org.jetbrains.jps.javac.APIWrappers$ProcessorWrapper.init(APIWrappers.java:105)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jetbrains.jps.javac.APIWrappers$1.invoke(APIWrappers.java:213)
	at com.sun.proxy.$Proxy29.init(Unknown Source)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$ProcessorState.<init>(JavacProcessingEnvironment.java:686)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.next(JavacProcessingEnvironment.java:786)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:881)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1222)
	at jdk.compiler/com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1334)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1258)
	at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:936)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
	... 21 more

They suggest using following code to unwrap JavacProcessingEnvironment

```
private static <T> T jbUnwrap(Class<? extends T> iface, T wrapper) {
    T unwrapped = null;
    try {
      final Class<?> apiWrappers = wrapper.getClass().getClassLoader().loadClass("org.jetbrains.jps.javac.APIWrappers");
      final Method unwrapMethod = apiWrappers.getDeclaredMethod("unwrap", Class.class, Object.class);
      unwrapped = iface.cast(unwrapMethod.invoke(null, iface, wrapper));
    }
    catch (Throwable ignored) {}
    return unwrapped != null? unwrapped : wrapper;
  }
```

Obviously, unwrapping should only be used in case processing environment is wrapped and would only work if it was wrapped by IntelliJ, not if it was wrapped by something else.

Is such modification acceptable for checkerframework?

**Comment by mernst** [comment]
Thank you for raising this issue.

The modification you suggest is acceptable.  I'm not sure exactly where to put it, and I would not be able to test it since I am not an IntelliJ user.  Could you please open a pull request?  Thanks!

**Comment by MichalStehlikCz** [comment]
Unwrap suggested by JetBrains does not work, but it was already fixed in Lombok in https://github.com/rzwitserloot/lombok/issues/2592, so I will try to replicate their approach (as it seems to work).

I also included unwrapping needed for gradle incremental compilation (#2401) as part of that issue is similar problem.

It would be better to only unwrap JavacProcessingEnvironment only where needed, but given that it is not clear which javac methods require JavacProcessingEnvironment (as for backward binary compatibility, they left ProcessingEnvironment in method signature and only retype inside methods) and that there are loads of places where Trees.instance is used (even though in most of them, it could be replaced with retrieving trees instance from SourceChecker), I unwrapped it at beginning of processor initialization.
